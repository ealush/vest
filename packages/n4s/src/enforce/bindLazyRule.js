import { RUN_RULE } from 'enforceKeywords';
import genRuleProxy from 'genRuleProxy';
import isFunction from 'isFunction';
import runtimeRules from 'runtimeRules';

// Initiates a chain of functions directly from the `enforce`
// function - that's even though we do not have any closure
// there to store that data.
export default function bindLazyRule(ruleName) {
  const registeredRules = [];

  const addFn = fnName => (...args) => {
    registeredRules.push(
      Object.defineProperty(
        value => runtimeRules[fnName](value, ...args),
        'name',
        { value: fnName }
      )
    );

    const returnvalue = genRuleProxy({}, addFn);

    return Object.assign(returnvalue, {
      [RUN_RULE]: getValue => {
        return registeredRules.every(fn => {
          // This  inversion of control when getting the value is
          // required in order to pass the function over to `shape`
          // so it can make the decision which args to pass to `optional`
          return fn(isFunction(getValue) ? getValue(fn.name) : getValue);
        });
      },
    });
  };

  return addFn(ruleName);
}
