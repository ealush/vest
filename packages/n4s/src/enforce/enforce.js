import compounds from 'compounds';
import { RUN_RULE } from 'enforceKeywords';
import runner from 'enforceRunner';
import isFunction from 'isFunction';
import isRule from 'isRule';
import proxySupported from 'proxySupported';
import rules from 'rules';

const rulesObject = Object.assign(rules(), compounds);

let rulesList = proxySupported() ? null : Object.keys(rulesObject);

const Enforce = value => {
  const proxy = genRuleProxy(enforce, ruleName => (...args) => {
    runner(rulesObject[ruleName], value, args);
    return proxy;
  });
  return proxy;
};

const enforce = genRuleProxy(Enforce, bindLazyRule);

enforce.extend = customRules => {
  Object.assign(rulesObject, customRules);

  if (!proxySupported()) {
    rulesList = Object.keys(rulesObject);
    genRuleProxy(Enforce, bindLazyRule);
  }

  return enforce;
};

export default enforce;

// Creates a proxy object that has access to all the rules
function genRuleProxy(target, output) {
  if (proxySupported()) {
    return new Proxy(target, {
      get: (target, fnName) => {
        if (isRule(rulesObject, fnName)) {
          return output(fnName);
        }

        return target[fnName];
      },
    });
  } else {
    /**
     * This method is REALLY not recommended as it is slow and iterates over
     * all the rules for each direct enforce reference. We only use it as a
     * lightweight alternative for the much faster proxy interface
     */
    return rulesList.reduce((target, fnName) => {
      return Object.defineProperties(target, {
        [fnName]: { get: () => output(fnName), configurable: true },
      });
    }, target);
  }
}

// Initiates a chain of functions directly from the `enforce`
// function - that's even though we do not have any closure
// there to store that data.
function bindLazyRule(ruleName) {
  const registeredRules = [];

  const addFn = fnName => (...args) => {
    registeredRules.push(
      Object.defineProperty(
        value => rulesObject[fnName](value, ...args),
        'name',
        { value: fnName }
      )
    );

    const returnvalue = genRuleProxy({}, addFn);

    return Object.assign(returnvalue, {
      [RUN_RULE]: getValue => {
        return registeredRules.every(fn => {
          // This  inversion of control when getting the value is
          // required in order to pass the function over to `shape`
          // so it can make the decision which args to pass to `optional`
          return fn(isFunction(getValue) ? getValue(fn.name) : getValue);
        });
      },
    });
  };

  return addFn(ruleName);
}
