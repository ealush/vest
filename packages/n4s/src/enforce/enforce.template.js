import runner from 'enforceRunner';
import genRuleProxy from 'genRuleProxy';
import isFunction from 'isFunction';
import runLazyRules from 'runLazyRules';
import runtimeRules from 'runtimeRules';
import withArgs from 'withArgs';

export default function bindTemplate(enforce) {
  enforce.template = withArgs(rule => {
    const template = value => {
      runner(runLazyRules.bind(null, rule), value);
      const proxy = genRuleProxy({}, ruleName =>
        withArgs(args => {
          runner(runtimeRules[ruleName], value, args);
          return proxy;
        })
      );
      return proxy;
    };

    template.test = getValue =>
      runLazyRules(rule, isFunction(getValue) ? getValue() : getValue);

    return template;
  });
}
