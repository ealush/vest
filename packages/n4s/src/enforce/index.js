import { isRule, proxySupported } from '../lib';
import rules from '../rules';
import runner from './runner';

const rulesObject = { ...rules };

let enforce, rulesList;

if (proxySupported()) {
  enforce = value => {
    const proxy = new Proxy(rulesObject, {
      get: (rules, fnName) => {
        if (!isRule(rules, fnName)) {
          return;
        }

        return (...args) => {
          runner(rules[fnName], value, ...args);
          return proxy;
        };
      },
    });
    return proxy;
  };
} else {
  rulesList = Object.keys(rulesObject);

  enforce = value =>
    rulesList.reduce(
      (allRules, fnName) =>
        Object.assign(allRules, {
          ...(isRule(rulesObject, fnName) && {
            [fnName]: (...args) => {
              runner(rulesObject[fnName], value, ...args);
              return allRules;
            },
          }),
        }),
      {}
    );
}

enforce.extend = customRules => {
  Object.assign(rulesObject, customRules);

  if (!proxySupported()) {
    rulesList = Object.keys(rulesObject);
  }
};

export default enforce;
