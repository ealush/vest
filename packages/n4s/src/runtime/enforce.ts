import assign from 'assign';

import eachEnforceRule from 'eachEnforceRule';
import { ctx, TEnforceContext } from 'enforceContext';
import enforceEager, { TEnforceEager } from 'enforceEager';
import genEnforceLazy, { TLazyRules } from 'genEnforceLazy';
import isProxySupported from 'isProxySupported';
import { TRule, KBaseRules, baseRules, getRule } from 'runtimeRules';
/**
 * Enforce is quite complicated, I want to explain it in detail.
 * It is dynamic in nature, so a lot of proxy objects are involved.
 *
 * Enforce has two main interfaces
 * 1. eager
 * 2. lazy
 *
 * The eager interface is the most commonly used, and the easier to understand.
 * It throws an error when a rule is not satisfied.
 * The eager interface is declared in enforceEager.ts and it is quite simple to understand.
 * enforce is called with a value, and the return value is a proxy object that points back to all the rules.
 * When a rule is called, the value is mapped as its first argument, and if the rule passes, the same
 * proxy object is returned. Otherwise, an error is thrown.
 *
 * The lazy interface works quite differently. It is declared in genEnforceLazy.ts.
 * Rather than calling enforce directly, the lazy interface has all the rules as "methods" (only by proxy).
 * Calling the first function in the chain will initialize an array of calls. It stores the different rule calls
 * and the parameters passed to them. None of the rules are called yet.
 * The rules are only invoked in sequence once either of these chained functions are called:
 * 1. test(value)
 * 2. run(value)
 *
 * Calling run or test will call all the rules in sequence, with the difference that test will only return a boolean value,
 * while run will return an object with the validation result and an optional message created by the rule.
 */

function genEnforce(): TEnforce {
  const target = {
    context: () => ctx.useX(),
    extend: (customRules: TRule) => {
      assign(baseRules, customRules);
    },
  } as TEnforce;

  if (!isProxySupported()) {
    eachEnforceRule((ruleName: KBaseRules) => {
      // Only on the first rule access - start the chain of calls
      target[ruleName] = genEnforceLazy(ruleName);
    });

    return assign(enforceEager, target);
  }

  return new Proxy(assign(enforceEager, target) as TEnforce, {
    get: (target: TEnforce, key: string) => {
      if (key in target) {
        return target[key];
      }

      if (!getRule(key)) {
        return;
      }

      // Only on the first rule access - start the chain of calls
      return genEnforceLazy(key);
    },
  });
}

export const enforce = genEnforce();

type TEnforce = TEnforceMethods & TLazyRules & TEnforceEager;

type TEnforceMethods = {
  context: () => TEnforceContext;
  extend: (customRules: TRule) => void;
};
