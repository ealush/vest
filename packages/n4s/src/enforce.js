import bindLazyRule from 'bindLazyRule';
import bindExtend from 'enforce.extend';
import bindTemplate from 'enforce.template';
import runner from 'enforceRunner';
import genRuleProxy from 'genRuleProxy';
import runtimeRules from 'runtimeRules';
import withArgs from 'withArgs';

const Enforce = value => {
  const proxy = genRuleProxy({}, ruleName =>
    withArgs(args => {
      runner(runtimeRules[ruleName], value, args);
      return proxy;
    })
  );
  return proxy;
};

const enforce = genRuleProxy(Enforce, bindLazyRule);

bindExtend(enforce, Enforce);
bindTemplate(enforce);

export default enforce;
