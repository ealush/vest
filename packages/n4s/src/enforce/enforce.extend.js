import bindLazyRule from 'bindLazyRule';
import genRuleProxy from 'genRuleProxy';
import proxySupported from 'proxySupported';
import runtimeRules from 'runtimeRules';

export default function bindExtend(enforce, Enforce) {
  enforce.extend = customRules => {
    Object.assign(runtimeRules, customRules);

    if (!proxySupported()) {
      genRuleProxy(Enforce, bindLazyRule);
    }

    return enforce;
  };
}
