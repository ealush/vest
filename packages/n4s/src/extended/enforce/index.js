import enforce from '../../enforce';
import rules, {
  isAlphanumericLocales,
  isMobilePhoneLocales,
  isPostalCodeLocales,
} from '../rules';

enforce.extend(rules);

export default Object.assign(enforce, {
  isAlphanumericLocales,
  isMobilePhoneLocales,
  isPostalCodeLocales,
});
