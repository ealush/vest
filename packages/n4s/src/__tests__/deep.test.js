import enforce from 'enforce';

describe('Deep enforcements', () => {
  it('Should match sucessful shapshot', () => {
    expect(
      enforce.shape(User).run({
        age: 18,
        country: 'US',
        user: {
          id: 20,
          name: { first: 'Teddie', last: 'Montes' },
        },
      })
    ).toMatchSnapshot();
  });

  it('Should match a partially failing shapshot', () => {
    expect(
      enforce.shape(User).run({
        age: 18,
        country: 'US',
        user: {
          id: null,
          name: { last: 'Montes' },
        },
      })
    ).toMatchSnapshot();
  });

  it('Should match a failing snapshot', () => {
    expect(enforce.shape(User).run({})).toMatchSnapshot();
    expect(enforce.shape(User)).not.toPassWith({});
  });
});

const User = {
  age: enforce.allOf(enforce.isNumeric(), enforce.greaterThanOrEquals(18)),
  country: enforce.oneOf(
    enforce.equals('US'),
    enforce.equals('Peru'),
    enforce.equals('Russia')
  ),
  user: enforce.shape({
    name: enforce.shape({
      first: enforce.isString().longerThan(2),
      last: enforce.isString().longerThan(1),
      middle: enforce.optional(enforce.isString()),
    }),
    id: enforce.anyOf(enforce.isString(), enforce.isNumber()),
  }),
};
