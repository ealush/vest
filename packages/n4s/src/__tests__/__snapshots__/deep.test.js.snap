// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deep enforcements Should match a failing snapshot 1`] = `
RuleResult {
  "children": Object {
    "age": RuleResult {
      "failed": true,
      "hasErrors": true,
      "hasWarnings": false,
      "warn": false,
    },
    "country": RuleResult {
      "failed": true,
      "hasErrors": true,
      "hasWarnings": false,
      "warn": false,
    },
    "user": RuleResult {
      "failed": true,
      "hasErrors": true,
    },
  },
  "failed": true,
  "hasErrors": true,
  "hasWarnings": false,
}
`;

exports[`Deep enforcements Should match a partially failing shapshot 1`] = `
RuleResult {
  "children": Object {
    "age": RuleResult {
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
      "warn": false,
    },
    "country": RuleResult {
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
      "warn": false,
    },
    "user": RuleResult {
      "children": Object {
        "id": RuleResult {
          "failed": true,
          "hasErrors": true,
          "hasWarnings": false,
          "warn": false,
        },
        "name": RuleResult {
          "children": Object {
            "first": RuleResult {
              "failed": true,
              "hasErrors": true,
              "hasWarnings": false,
              "warn": false,
            },
            "last": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "warn": false,
            },
            "middle": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "warn": false,
            },
          },
          "failed": true,
          "hasErrors": true,
          "hasWarnings": false,
        },
      },
      "failed": true,
      "hasErrors": true,
      "hasWarnings": false,
    },
  },
  "failed": true,
  "hasErrors": true,
  "hasWarnings": false,
}
`;

exports[`Deep enforcements Should match sucessful shapshot 1`] = `
RuleResult {
  "children": Object {
    "age": RuleResult {
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
      "warn": false,
    },
    "country": RuleResult {
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
      "warn": false,
    },
    "user": RuleResult {
      "children": Object {
        "id": RuleResult {
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "warn": false,
        },
        "name": RuleResult {
          "children": Object {
            "first": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "warn": false,
            },
            "last": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "warn": false,
            },
            "middle": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "warn": false,
            },
          },
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
        },
      },
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
    },
  },
  "failed": false,
  "hasErrors": false,
  "hasWarnings": false,
}
`;
